### DISCUSSION TITLE
Using ProtectedData in MSBuild

### DISCUSSION THREAD

**Discussion by shadow-cs** [dotnet/msbuild#7420] [Q&A]
A while back (.NET Core 3.1) we implemented a custom task that used `ProtectedData` API using the `RoslynCodeTaskFactory` to store authentication tokens needed during the build and it worked just fine. Then we switched to a different approach so this wasn't tested for a while.

Then on a different project we had that need again so we pulled the previously working solution out of the hat only to figure out it doesn't work in Visual Studio 2022 (CLI build using `dotnet build` works just fine). I haven't tested VS 2019 as we rely on .NET 6.

I cannot shake the feeling this is somehow related to <https://github.com/dotnet/sdk/issues/23498#issuecomment-1050119918> since there clearly is an assembly loading problem somewhere.

I wasn't able to work around this so we basically copy-pasted `ProtectedData` from the runtime (it's only a few lines of code). So we don't have any issue right now but maybe solving this could help others :wink:.

Could you help me @baronfel get to the bottom of this?

Thanks in advance

**Comment by rainersigwald** [comment]
This is very likely the same root cause as that issue and so should be resolved by #7415.

**Reply by shadow-cs** [reply]
Thanks, I'll try to keep this in mind and confirm this as soon as the new preview of VS with these changes is out.

**Reply by rainersigwald** [reply]
Ah, I'm sorry, I misread your situation. I think you might actually be seeing dotnet/msbuild#5106, though you indicated that it worked before, which it shouldn't have . . .
